libplatsupport/src/mach/eswin/ltimer.c

```diff
@@ -0,0 +1,239 @@
+        if (time >= ns) {
+            ZF_LOGE("Requested time %"PRIu64" earlier than current time %"PRIu64, ns, time);
+            return ETIME;
+        }
+        return eswin_timer_set_timeout(&timers->timer[TIMEOUT_TIMER], ns - time, false);
+    }
+    case TIMEOUT_RELATIVE:
+        return eswin_timer_set_timeout(&timers->timer[TIMEOUT_TIMER], ns, false);
+    case TIMEOUT_PERIODIC:
+        return eswin_timer_set_timeout(&timers->timer[TIMEOUT_TIMER], ns, true);
+    }
+
+    return EINVAL;
+}
+
+static int reset(void *data)
+{
+    assert(data);
+    eswin_ltimer_t *timers = (eswin_ltimer_t *)data;
+
+    eswin_timer_disable(&timers->timer[COUNTER_TIMER]);
+    eswin_timer_reset(&timers->timer[COUNTER_TIMER]);
+    eswin_timer_enable(&timers->timer[COUNTER_TIMER]);
+
+    eswin_timer_disable(&timers->timer[TIMEOUT_TIMER]);
+    eswin_timer_reset(&timers->timer[TIMEOUT_TIMER]);
+
+    return 0;
+}
+
+static void destroy(void *data)
+{
+    assert(data);
+    eswin_ltimer_t *timers = (eswin_ltimer_t *)data;
+    int error;
+
+    eswin_timer_disable(&timers->timer[COUNTER_TIMER]);
+    eswin_timer_disable(&timers->timer[TIMEOUT_TIMER]);
+
+    ps_pmem_unmap(&timers->ops, pmems[0], timers->vaddr);
+
+    error = ps_irq_unregister(&timers->ops.irq_ops, timers->irq_id[COUNTER_TIMER]);
+    ZF_LOGE_IF(error, "Failed to uregister counter timer IRQ");
+
+    error = ps_irq_unregister(&timers->ops.irq_ops, timers->irq_id[TIMEOUT_TIMER]);
+    ZF_LOGE_IF(error, "Failed to uregister timeout timer IRQ");
+
+    error = ps_free(&timers->ops.malloc_ops, sizeof(*timers), timers);
+    ZF_LOGE_IF(error, "Failed to free device struct memory");
+}
+
+int ltimer_default_init(ltimer_t *ltimer,
+                        ps_io_ops_t ops,
+                        ltimer_callback_fn_t callback,
+                        void *callback_token)
+{
+    assert(ltimer);
+
+    ltimer->get_num_irqs = get_num_irqs;
+    ltimer->get_nth_irq = get_nth_irq;
+    ltimer->get_num_pmems = get_num_pmems;
+    ltimer->get_nth_pmem = get_nth_pmem;
+    ltimer->get_time = get_time;
+    ltimer->set_timeout = set_timeout;
+    ltimer->reset = reset;
+    ltimer->destroy = destroy;
+
+    int error = ps_calloc(&ops.malloc_ops, 1, sizeof(eswin_ltimer_t), &ltimer->data);
+    if (error) {
+        ZF_LOGE("Memory allocation failed with error %d", error);
+        return error;
+    }
+
+    eswin_ltimer_t *timers = ltimer->data;
+
+    timers->ops = ops;
+    timers->user_callback = callback;
+    timers->user_callback_token = callback_token;
+
+    timers->callback_data.ltimer = ltimer;
+    timers->callback_data.irq = &irqs[0];
+    timers->callback_data.irq_handler = ltimer_handle_irq;
+
+    irq_id_t irq_id = ps_irq_register(&ops.irq_ops,
+                                      irqs[0],
+                                      handle_irq_wrapper,
+                                      &timers->callback_data);
```

Comment:
I really don't understand the point of `handle_irq_wrapper`, but everyone is using it.